### PR TITLE
Custom size mode

### PR DESCRIPTION
**Pull request by jd-carroll**
First cut of allowing custom size modes.

### CODE REVIEW

File: core/Size.js [hunk 1]
```diff
@@ -113,21 +114,46 @@ function setVec (vec, x, y, z) {
  * @method
  *
  * @param {String|Number} val The Size mode to resolve.
+ * @param {boolean} quiet Whether to throw an error for invalid mode
  *
  * @return {Number} the resolved size mode from the enumeration.
  */
-function resolveSizeMode (val) {
+function resolveSizeMode (val, quiet) {
```
Comment (alexanderGugel) [inline core/Size.js]: Why is this option needed?

Comment (jd-carroll) [inline core/Size.js]: `resolveSizeMode` is called while iterating through the list of input params for the customized function.  If it returns a value, then we know we need to retrieve the value for one of `absolute`, `relative`, or `render`.  If it doesn't return a value, then we use the value from the array as is.

File: core/Node.js [hunk 2]
```diff
@@ -1053,15 +1053,25 @@ Node.prototype.setOpacity = function setOpacity (val) {
...
+ *                                 size.
+ * @param {Array} customParams     An array of parameters to pass into the 
+ *                                 custom calculation.  Updates to the original
+ *                                 array will be reflected in the next 
+ *                                 calculation.
  *
  * @return {Node} this
  */
-Node.prototype.setSizeMode = function setSizeMode (x, y, z) {
+Node.prototype.setSizeMode = function setSizeMode (x, y, z, customMode, customParams) {
```
Comment (jd-carroll) [inline core/Node.js]: I'm not real satisfied with this... 5 input params is a lot so I am definitely open to other suggestions

File: core/test/Size.js [hunk 3]
```diff
@@ -45,7 +45,15 @@ test('Size', function(t) {
...
-        // TODO
+        // TODO: Size.DEFAULT
+        
+        // TODO: Size.ABSOLUTE
+        
+        // TODO: Size.RELATIVE
+        
+        // TODO: Size.RENDER
+        
+        // TODO: Size.CUSTOM
```
Comment (jd-carroll) [inline core/test/Size.js]: If someone could share a general framework / thought process for testing the different modes, I can take a stab at writing the tests.  We definitely need these as there was already one issue #354 & #355

File: core/Size.js (hunk 1 above)
Comment (jd-carroll) [inline core/Size.js]: ref: https://github.com/Famous/engine/pull/373/files#diff-7ef39360caa0712ad34bf606081c9ea7R363

**Comment by jd-carroll** [comment]
I am now utilizing this code in my local repo, if this could have a final review that would be appreciated.  
May need to tweak a couple of things:
- Passing fn + params into `Size` component ([L1065](https://github.com/Famous/engine/pull/373/files#diff-20164165518dcfb14298730df6349123R1065))
- Creating new array for resolving dependencies ([L360](https://github.com/Famous/engine/pull/373/files#diff-7ef39360caa0712ad34bf606081c9ea7R360))
- Scope of custom mode fn ([L370](https://github.com/Famous/engine/pull/373/files#diff-7ef39360caa0712ad34bf606081c9ea7R370))

@alexanderGugel

**Comment by michaelobriena** [comment]
@jd-carroll I am a bit torn on this.

While I think this is really cool and adds flexibility I wonder if custom size mode functionality should be part of the Node or if it should just be a component attached to a Node that manages the size.  I think the counter would be that Relative and Render size should then be be components as well.

Anyone else have thoughts on this?  Need to think on this for a bit.

**Comment by michaelobriena** [comment]
@jd-carroll either way, thanks for introducing the idea!

**Comment by alexanderGugel** [comment]
After a lot of back and forth on this, I'm against this change. At least at this point in time. I agree with @michaelobriena that this should be a component. In fact, I would even argue that **every** size mode should be a component and pixel size should be the *only* core size mode.

After all, I don't think this is something we currently want/ need and it introduces unnecessary complexity.

@jd-carroll Thanks for the PR though, but I don't think this is something we currently want/ need.

@michaelobriena Can we close this?

**Pull request closed without merging**